Make the quorum generator & Validator routines @safe and nothrow

AndrejMitrovic opened the pull request.

AndrejMitrovic committed:
Make most nominating routines @safe nothrow

AndrejMitrovic committed:
Rename verifyBuildSCPConfig to buildSCPConfig

Remove validation from the renamed routine
because the quorums are no longer user-configurable,
but generated by the quorum generator which
has tests for quorum validity.

AndrejMitrovic committed:
Change quorum threshold type to uint

The type was already being converted to uint
when being passed to SCP, there's no need to
keep it architecture-dependent.

This also allows making the conversion
routines nothrow

Geod24 commented:
> For whatever reason the autogenerated opAssign [...]

Maybe because the dtor is `@system` ?

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -122,23 +122,23 @@ extern(D):
         foreach (qc; other_quorums)
         {
-            auto quorum_set = verifyBuildSCPConfig(qc);
+            auto quorum_set = buildSCPConfig(qc);
```
Would it be less confusing to use a different name from the `this.quorum_set`?

AndrejMitrovic replied:
Yes, in fact the field is badly named. I will add another commit.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -122,23 +122,23 @@ extern(D):
-        auto quorum_set = verifyBuildSCPConfig(quorum);
-        this.scp.updateLocalQuorumSet(quorum_set);
-        auto shared_set = makeSharedSCPQuorumSet(this.scp.getLocalQuorumSet());
+        auto quorum_set = buildSCPConfig(quorum);
```
Same question for the name for the variable.

AndrejMitrovic committed:
Rename quorum_set to known_quorums

The field is a map of all known quorum sets,
not just the node's own quorum set.

AndrejMitrovic commented:
Added another commit renaming the field as suggested.

AndrejMitrovic commented:
> Maybe because the dtor is @system ?

Yes that was it. I wish the diagnostic was better.

AndrejMitrovic commented:
https://issues.dlang.org/show_bug.cgi?id=21043

MichaelKim20 commented on source/scpd/scp/QuorumSetUtils.d:
```diff
@@ -22,4 +22,5 @@ bool isQuorumSetSane(ref const SCPQuorumSet qSet, bool extraChecks,
 
 // normalize the quorum set, optionally removing idToRemove
-void normalizeQSet(ref SCPQuorumSet qSet, const(NodeID)* idToRemove = null);
+void normalizeQSet(ref SCPQuorumSet qSet, const(NodeID)* idToRemove = null)
+    nothrow @nogc;
```
There's something I don't understand, so let me ask you.
Use 'try catch' at C++ to modify it.
Or should we pack this function once more and make an exception there?

AndrejMitrovic replied:
This routine does not throw any exceptions.

MichaelKim20 replied:
I see.

MichaelKim20 commented on source/scpd/types/Utils.d:
```diff
@@ -19,7 +19,7 @@ import scpd.types.Stellar_types;
 
 extern(C++) public shared_ptr!SCPQuorumSet makeSharedSCPQuorumSet (
-    ref const(SCPQuorumSet));
+    ref const(SCPQuorumSet)) nothrow @nogc @safe;
```
I have the same question as above.
There's something I don't understand, so let me ask you.
Add exception processing at C++.
Or should we pack this function once more and make an exception there?

AndrejMitrovic replied:
The routine does not throw. We're passing a reference and it's calling `std::make_shared`, it won't fail.

MichaelKim20 reviewed (commented):
Very good. I have two questions.

Geod24 commented:
Still a draft ?

AndrejMitrovic commented:
No, just infinite wait bug.

AndrejMitrovic commented:
I'll rebase.

MichaelKim20 reviewed (approved):
LGTM

AndrejMitrovic commented:
@Geod24 do you want to look into it more, or it looks good to you?

AndrejMitrovic commented:
If you want me to squash some stuff let me know. 👋

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -87,11 +87,18 @@ public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
     auto RNG_gen = getGenerator(key);
     auto stake_amounts = stakes.map!(stake => stake.amount.integral);
 
+    auto assumeNothrow (T)(lazy T exp) nothrow
```
```suggestion
    static auto assumeNothrow (T)(lazy T exp) nothrow
```

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -70,7 +70,7 @@ private enum double THRESHOLD = 0.80;
 
 public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
-    in Hash[] utxo_keys, UTXOFinder finder )
+    in Hash[] utxo_keys, UTXOFinder finder ) @safe nothrow
```
```suggestion
    in Hash[] utxo_keys, scope UTXOFinder finder ) @safe nothrow
```

Geod24 reviewed (approved):
Just two nits, otherwise LGTM. Merge at will. I'd say squash the 3 commits that add `@safe` here and there if you want, but I'm okay either way.
